### PR TITLE
Fix: don't require NodeId if resolve_id is overwritten

### PR DESCRIPTION
**Pull request by devkral**
If resolve_id is overwritten it is valid that no NodeId annotation is found (custom logic). Don't fail in this case

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #2816

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Alexander for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/relay/types.py [hunk 1]
```diff
@@ -375,6 +374,11 @@ def __init_subclass__(cls, **kwargs: Any):
         ]
 
         if len(candidates) == 0:
+            # resolve_id is overwritten, allow unset _id_attr
+            for mro_cls in cls.__mro__:
+                if mro_cls != Node:
+                    if "resolve_id" in cls.__dict__:
+                        return
```
Comment (bellini666) [inline strawberry/relay/types.py:381]: nit: compare with `is` and avoid nesting ifs

```suggestion
                if mro_cls is not Node and "resolve_id" in cls.__dict__:
                    return
```

File: tests/relay/test_exceptions.py [hunk 2]
```diff
@@ -26,6 +26,16 @@ class Fruit(relay.Node):
...
 
 
+def test_not_raises_error_on_missing_node_id_annotation_but_resolve_id():
+    @strawberry.type
+    class Fruit(relay.Node):
+        code: str
+
+        @classmethod
+        def resolve_id(cls, root) -> str:
+            return "test"
```
Comment (bellini666) [inline tests/relay/test_exceptions.py:36]: Some things here:

1) This doesn't raise an exception, so probably the best place to put this is on relay tests itself. Maybe on `relay/test_types.py`?

2) Just to make it more complete, it would be nice to create a schema here and compare the it. Search for `dedent` on the code and you will find some tests that do that as an example

**Review by bellini666 — CHANGES_REQUESTED** [review]

**Comment by bellini666** [comment]
Also, don't forget to add a `RELEASE.md` file, just like [botberry suggested](https://github.com/strawberry-graphql/strawberry/pull/2834#issuecomment-1585414519), so it can be released once it is merged :)

File: tests/relay/test_exceptions.py (hunk 2 above)
Comment (devkral) [reply tests/relay/test_exceptions.py]: thanks for pointing to the right file

File: strawberry/relay/types.py (hunk 1 above)
Comment (devkral) [reply strawberry/relay/types.py]: done, and edge case solved

File: tests/relay/test_types.py [hunk 3]
```diff
@@ -148,3 +149,13 @@ class Foo:
     gid = relay.GlobalID(type_name="FruitAsync", node_id="1")
     with pytest.raises(TypeError):
         fruit = await gid.resolve_node(fake_info, ensure_type=Foo)
+
+
+def test_not_raises_error_on_missing_node_id_annotation_but_resolve_id():
```
Comment (bellini666) [inline tests/relay/test_types.py:154]: Just need to add a schema print test in here, as I mentioned in the previous review, to make sure that it prints correctly :)

**Review by bellini666 — CHANGES_REQUESTED** [review]

**Comment by bellini666** [comment]
@devkral btw, I'm migrating strawberry-django-plus to use the official relay implementation in this PR: https://github.com/blb-ventures/strawberry-django-plus/pull/235

I had to open https://github.com/strawberry-graphql/strawberry/pull/2838 to fix an issue that I found there. Don't know if that change would supersede this PR, but probably it will need to be adjusted to consider the new behaviour

**Comment by devkral** [comment]
yes this PR will supersede this PR. But maybe add the tests from this PR (plus one for test printer)

**Comment by devkral** [comment]
> yes this PR will supersede this PR. But maybe add the tests from this PR (plus one for test printer)

sry, was a bit in hurry ;). I will port the tests.

**Comment by devkral** [comment]
ported the tests and fixed a remaining issue
